### DISCUSSION TITLE
New loader request megathread

### DISCUSSION THREAD

**Discussion by triphora** [modrinth/.github#106] [Feature requests]
Please use the following template for new loader suggestions:
```
**What loader do you suggest?:** Loader name

**Please provide a link to the loader (website, GitHub, etc.):** https:// ...

**When, approximately, was this loader created?:** January 1970

**What project type(s) would this loader be applicable to?:** Mods, plugins, data packs, shaders, resource packs, modpacks

**Please provide links to at least three projects that use this loader:**
- https://github.com/ ...
- https://modrinth.com/ ...
- https://example.com/ ...

**Any additional context?:** You can remove this question if you want.
```

Here's an example, using an already existing Modrinth loader:

<details><summary>Example</summary>

**What loader do you suggest?:** Folia

**Please provide a link to the loader (website, GitHub, etc.):** https://github.com/PaperMC/Folia

**When, approximately, was this loader created?:** Early 2023

**What project type(s) would this loader be applicable to?:** Plugins

**Please provide links to at least three projects that use this loader:**
- https://modrinth.com/plugin/simple-voice-chat
- https://modrinth.com/plugin/plasmo-voice
- https://modrinth.com/plugin/freedomchat

</details>

Feel free to upvote individual requests and use comment threads. Please do not create a new thread if there is already an existing thread for your suggested loader.

Please note that we may require multiple upvotes before we consider adding a new loader.

**See also:** [the category request megathread](https://github.com/orgs/modrinth/discussions/105)

**Comment by ThexXTURBOXx** [comment]
**What loader do you suggest?:** ModLoaderMP (MLMP)

**Please provide a link to the loader (website, GitHub, etc.):** https://mcarchive.net/mods/modloadermp, originally at http://www.minecraftforum.net/topic/86765-166-sdks-mods-v5-more-bug-fixes/

**When, approximately, was this loader created?:** November 2010

**What project type(s) would this loader be applicable to?:** Mods, Modpacks

**Please provide links to at least three projects that use this loader:**
- SDK's Mods: https://www.minecraftforum.net/forums/mapping-and-modding-java-edition/minecraft-mods/1272343-1-3-2-sdks-mods-v1-aug-18th-modloadermp-updated
- Flan's Mod: https://www.minecraftforum.net/forums/mapping-and-modding-java-edition/minecraft-mods/1272467-flans-mod-5-5-2-update-1-12-2-100s-of-new-skins
- Block Helper: https://modrinth.com/mod/block-helper

**Any additional context?:** This loader is almost as dated as `ModLoader` itself. Also, when installing MLMP, make sure to install ModLoader first!

**Reply by Prospector** [reply]
Not really a mod loader as it's a library for Risugami's ModLoader. Has been uploaded on the Modrinth Archives: https://modrinth.com/mod/modloadermp

**Comment by ThexXTURBOXx** [comment]
**What loader do you suggest?:** Meddle

**Please provide a link to the loader (website, GitHub, etc.):** https://github.com/FyberOptic/Meddle

**When, approximately, was this loader created?:** July 2015

**What project type(s) would this loader be applicable to?:** Mods, _maybe_ Modpacks

**Please provide links to at least three projects that use this loader:**
- https://github.com/FyberOptic/MeddleMods
- https://github.com/FyberOptic/JustBackpack
- Many more found in this mega thread: https://www.minecraftforum.net/forums/mapping-and-modding-java-edition/minecraft-mods/2488387-meddle-minecraft-tweakclass-mod-loader-1-9-1-10-1#mods

**Any additional context?:** Allows to play with the same mods on 1.9 and 1.10 snapshots at the same time

**Comment by ThexXTURBOXx** [comment]
**What loader do you suggest?:** Jar-Modding

**Please provide a link to the loader (website, GitHub, etc.):** N/A

**When, approximately, was this loader created?:** May-June 2009 (right after the birth of Minecraft basically, maybe a month after)

**What project type(s) would this loader be applicable to?:** Mods, _maybe_ Modpacks

**Please provide links to at least three projects that use this loader:**
- https://optifine.net/downloads (at least old versions that are not anymore listed on the website, but still available under their original link: https://www.minecraftforum.net/forums/mapping-and-modding-java-edition/minecraft-mods/1286605-b1-4-1-9-optifine-history )
- https://www.minecraftforum.net/forums/mapping-and-modding-java-edition/minecraft-mods/1286604-shaders-mod-updated-by-karyonix
- https://mcarchive.net/mods/achievementpages
- https://modrinth.com/mod/class-loader-fixer
- https://modrinth.com/mod/font-fixer

**Any additional context?:** I know that it's a far stretch to call this a "loader", but yeah... That's how many old mods are installed: Extracting files into the `minecraft.jar` (and deleting the `META-INF` folder)

**Reply by IMB11** [reply]
The problem with jar modding is that most jar mods are not in patch format and instead just directly redistributed the game with the modified code inside of it. At least this was the case for optifine and a few other large jar mods.

**Reply by ThexXTURBOXx** [reply]
@mineblock11 Sadly, that is completely true. There is no convention on how to package jar mods. Some people just put the class files in a zip/jar (which would be the best way of course).
However, there are many other ways. For instance, the achievement pages mod from my post has a sub folder "bin" which contains the class files. And of course, there are many more esoteric ones who have zips inside zips etc.

**Comment by ThexXTURBOXx** [comment]
**What loader do you suggest?:** MineCraftPortCentral+ (MCPC+)

**Please provide a link to the loader (website, GitHub, etc.):** https://web.archive.org/web/20130708021047/http://www.mcportcentral.co.za/
For the Wiki, you can also use my mirror: https://femtopedia.de:8080/mcpc-wiki/index.php?title=MCPC_Ported_Mods

**When, approximately, was this loader created?:** June 2011

**What project type(s) would this loader be applicable to?:** Mods, Plugins, Modpacks

**Please provide links to at least three projects that use this loader:**
- https://modrinth.com/mod/block-helper
- All of these: https://femtopedia.de:8080/mcpc-wiki/index.php?title=Ports_for_1.2.3
- Or these: https://femtopedia.de:8080/mcpc-wiki/index.php?title=Ports_for_1.1
- Or even better, these: https://femtopedia.de:8080/mcpc-wiki/index.php?title=MCPC_Ported_Mods

**Any additional context?:** Nowadays, MCPC+ is also known as Cauldron, KCauldron and many other projects. Before MCPC+ existed, there was also krnlyng's MLMP+Bukkit project ([original thread](https://web.archive.org/web/20110811063424/https://www.minecraftforum.net/topic/382599-173-modloadermp-for-craftbukkit-and-canary/)) which eventually became the basis for MCPC+. Hence, it is most typically also considered "MCPC+".
I took it upon myself to archive as many builds of MCPC+ as possible: https://archive.org/details/mcpc_20220726
Also, Cauldron and KCauldron don't need extra versions of mods, the "usual" build also works. And only MCPC+ up to Minecraft 1.2.5 need special mod versions to work on it, for Minecraft versions up to 1.4.7, only some mods need special builds

**Comment by bangetto** [comment]
**What loader do you suggest?:** Optifine

**Please provide a link to the loader (website, GitHub, etc.):** https://optifine.net

**When, approximately, was this loader created?:** around minecraft 1.5

**What project type(s) would this loader be applicable to?:** Resource packs

**Please provide links to at least three projects that use this loader:**
- https://modrinth.com/resourcepack/what-painting-im-looking-at
- https://modrinth.com/resourcepack/variantarmory
- https://modrinth.com/resourcepack/visual_enchantments

**Any additional context?:** Like 1/4 of resources uses optifine in some way

**Reply by Minenash** [reply]
It's worth noting that in the future you'll be able to search by dependencies (which a mod would be). I don't think a mod that extends ResourcePacks will be given it's own loader

**Reply by bangetto** [reply]
The problem that most of the peoples don't add decencies to resources pack as a decency, they just mention it in the desc, and btw there's a lot of packs that literally don't work without optifine [like the ones mentioned above].

**Reply by ThexXTURBOXx** [reply]
OptiFine is not a `loader`. Such as `jar modding` also involves no loader. However, I don't think, it makes much sense to declare OptiFine as loader then, but rather as @Minenash suggested as dependency

**Reply by Minenash** [reply]
> The problem that most of the peoples don't add decencies to resources pack as a decency

I don't think you can have external dependencies at the moment. But I know it's a planned thing. But also authors would still have to mark something (either as a dependency or as a loader), so putting it as a launcher doesn't save much effort.

There's also the problem that for each mod that extends resource packs, that would be a new loader. Dependencies are a lot more flexible. And for any launcher using the api, they can grab the dependency. If it was a loader, then they would probably need to hard code every single mod that extends resource packs

**Comment by bangetto** [comment]
**What loader do you suggest?:** Chime

**Please provide a link to the loader (website, GitHub, etc.):** https://github.com/emilyploszaj/chime

**When, approximately, was this loader created?:** Dec 26, 2020

**What project type(s) would this loader be applicable to?:** Resource packs

**Please provide links to at least three projects that use this loader:**
- https://modrinth.com/resourcepack/what-painting-im-looking-at
- https://modrinth.com/resourcepack/variantarmory
- https://modrinth.com/resourcepack/visual_enchantments

**Reply by IMB11** [reply]
> Chime is a mod

Mark it as a dependency, not a loader.

**Comment by ThatGravyBoat** [comment]
**What loader do you suggest?:** Minestom

**Please provide a link to the loader (website, GitHub, etc.):** https://minestom.net/ 

**When, approximately, was this loader created?:** August 2019

**What project type(s) would this loader be applicable to?:** Plugins

**Please provide links to at least three projects that use this loader:**
- https://github.com/emortalmc/TNT
- https://github.com/Cubxity/UnifiedMetrics / https://modrinth.com/plugin/unifiedmetrics
- https://github.com/dev-hydrogen/Nightclub

**Any additional context?:** 
You can find extensions which are what they call plugins on their site https://minestom.net/extensions/
This was requested in #128 but was closed as it was thought that Minestom was another game

**Reply by samuelh2005** [reply]
Minestom no longer have an extension system, they have become a library rather then a thing that does everything.

**Comment by triphora** [comment]
**What loader do you suggest?:** Java agent

**What project type(s) would this loader be applicable to?:** Mods, plugins

**Reply by ThexXTURBOXx** [reply]
Two example mods:
- https://modrinth.com/mod/pipeblocker
- https://modrinth.com/mod/serializationisbad

**Comment by Vetpetmon** [comment]
**What loader do you suggest?:** Cleanroom

**Please provide a link to the loader (website, GitHub, etc.):** [GitHub](https://github.com/CleanroomMC/Cleanroom/tree/dev/mixin3), [Website](https://repo.cleanroommc.com/#/)

**When, approximately, was this loader created?:** February 2023

**What project type(s) would this loader be applicable to?:** Mods, modpacks

**Please provide links to at least three projects that use this loader:**
- https://modrinth.com/mod/wyrms (Compatible out of the box, software support for Cleanroom loader/launcher is now endorsed)
- https://modrinth.com/mod/geckolib (Compatible out of the box as well)
- https://modrinth.com/mod/the-impossible-library (Also compatible)
- https://modrinth.com/mod/scapeandrunparasites (Big name mod, compatible with Cleanroom)
- https://modrinth.com/mod/mixinbooter (Included in the loader)
- https://www.curseforge.com/minecraft/mc-mods/coroutil/files/2902920 (And any of Coro's 1.12.2 mods will work. Coro's 1.12.2 mods aren't on Modrinth, but they can run under this loader with minimal issues)


**Any additional context?:** 

This is a fork/continuation of the 1.12.2 Forge Mod Loader that is experimental (no official MDK yet), but allows for that MC version to be ran under Java 21, comes with the latest Mixin version, and latest (customized) rendering libraries. It's different enough to be distinguished from Forge. Many projects aren't specifically made for Cleanroom, but are still compatible due to the patching and classloading performed by the loader; some projects perform better, and benefit from Mixinbooter being included in the loader itself, a way to distinguish Cleanroom-compatible/supported mods would be very beneficial for users. 

It's been noted that some mods, like RandomPatches, Better With Mods, many mods other mods that may require older versions of Mixin or the original rendering library used by Forge, and so on. Some mods are not forwards-compatible with Cleanroom, so distinctions should be made, and we'll eventually see forks or patches for old/abandoned 1.12.2 projects to become compatible with Cleanroom.

**Comment by SpaceWalkerRS** [comment]
**What loader do you suggest?:** Fabric/Quilt offshoots for Legacy Minecraft versions
- OrnitheMC (MC indev up to 1.14.4)
- Legacy Fabric (MC 1.3 up to 1.13.2)
- Minecraft Cursed Legacy/Babric (MC b1.7.3 and 1.2.5)

**Please provide a link to the loader (website, GitHub, etc.):**
- OrnitheMC: https://ornithemc.net
- Legacy Fabric: https://legacyfabric.net
- Minecraft Cursed Legacy: https://minecraft-cursed-legacy.github.io
- Babric: https://babric.github.io

**When, approximately, was this loader created?:**
- OrnitheMC: December 2021
- Legacy Fabric: February 2020(?)
- Minecraft Cursed Legacy: July 2019(?)
- Babric: 2022(?)

**What project type(s) would this loader be applicable to?:**
- mods
- modpacks

**Please provide links to at least three projects that use this loader:**
- OrnitheMC: 
  - Legacy Optifabric: https://modrinth.com/mod/legacy-optifabric
  - Legacy-lwjgl3: https://modrinth.com/mod/moehreag-legacy-lwjgl3
  - Mod Menu Ornithe: https://modrinth.com/mod/modmenu-ornithe
- Legacy Fabric:
  - Legacy Fabric API: https://modrinth.com/mod/legacy-fabric-api
  - FPS Plus Legacy Fabric: https://modrinth.com/mod/fpspluslf
  - Legacy Mod Menu: https://modrinth.com/mod/legacy-mod-menu
- Minecraft Cursed Legacy/Babric:
  - Cursed Legacy API: https://modrinth.com/mod/cursed-legacy-api
  - OptiBabric: https://modrinth.com/mod/optibabric
  - Accessory API: https://modrinth.com/mod/accessory-api

**Any additional context?:**
Neither Fabric nor Quilt officially support Minecraft versions 1.13.2 and older (and even support for 1.14 snapshots is shaky). However, Fabric Loader was always designed to be "version agnostic" and the developers have shown to be willing to fix bugs in order to allow these older versions to be run with Fabric. This has led to several offshoots in order to provide Intermediaries, mappings, APIs, and tooling to make Fabric modding on these "Legacy" Minecraft versions possible.

In reality, none of these are new/different mod loaders. Both Ornithe and Legacy Fabric use stock Fabric and/or Quilt Loader, whereas Cursed Legacy used a lightly modified Fabric Loader to support merging the client and server jars (this needs extra work in all versions prior to 1.3).

So far mod developers have been getting by on the Modrinth website by selecting "Fabric" and/or "Quilt" as the mod loader for their Legacy Minecraft mods. But this is not ideal, as it does not distinguish between e.g. Ornithe and Legacy Fabric (which are *not* inter-compatible). Ideally these projects would be added as a separate mod loader option, or perhaps as a sub option under Fabric and/or Quilt. Then mod developers could select the proper mod loader for their mods, and users could more easily find mods for these projects.

The situation with the Modrinth app is more dire. As far as I know, most of these Legacy modding projects have some kind of "MultiMC/Prism instance generator", but there is no such thing for the Modrinth Launcher. This means there is currently no (easy) way to install them on the Modrinth app. For Ornithe and Legacy Fabric this mostly comes down to grabbing the right Intermediary, but for Cursed Legacy/Babric the modified loader must be installed as well.

**Reply by ThexXTURBOXx** [reply]
I second this. My mod [Block Helper](https://modrinth.com/mod/block-helper) is also supported on Babric/Cursed Legacy (through [Apron](https://github.com/FabricCompatibilityLayers/Apron)) and I just marked it as "Fabric". Using another label for that would be better for the users' experience when searching for such mods.

**Reply by ShrikerChestTrees** [reply]
> I second this. My mod [Block Helper](https://modrinth.com/mod/block-helper) is also supported on Babric/Cursed Legacy (through [Apron](https://github.com/FabricCompatibilityLayers/Apron)) and I just marked it as "Fabric". Using another label for that would be better for the users' experience when searching for such mods.

For modpacks is there a lifehack?

**Comment by ShrikerChestTrees** [comment]
**What loader do you suggest?:** Babric

**Please provide a link to the loader (website, GitHub, etc.):** [Babric](https://babric.github.io/)

**When, approximately, was this loader created?:** 7-10 March 2022

**What project type(s) would this loader be applicable to?:** Mods & modpacks

**Please provide links to at least three projects that use this loader:**
- [BetterAI BTA - github](https://github.com/baguchan/BetterAI-BTA)
- [OptiBabric - Modrinth](https://modrinth.com/mod/optibabric)
- [Macula - Modrinth](https://modrinth.com/mod/macula)

**Any additional context?:**
Those mods on Modrinth specify modloader not Babric but Fabric.
And modpacks can't be created at all due to the lack of such modloader.

**Reply by blryface** [reply]
babric was already suggested

**Reply by DanyGames2014** [reply]
> babric was already suggested

And same as every other loader here, completely ignored 🙂

**Comment by unilock** [comment]
**What loader do you suggest?:** NilLoader

**Please provide a link to the loader (website, GitHub, etc.):** https://git.sleeping.town/Nil/NilLoader

**When, approximately, was this loader created?:** May 2022

**What project type(s) would this loader be applicable to?:** Mods, modpacks

**Please provide links to at least three projects that use this loader:**
- The flagship: https://git.sleeping.town/Rewind/Upsilon
- https://github.com/LemmaEOF/aleph
- https://github.com/BetterThanUpdates/BabricNil
- https://github.com/Kneelawk/ClassicConnectToServer
- https://git.sleeping.town/Nil/DataFunk
- https://github.com/unilock/DecorateLater
- https://github.com/kumquat-ir/demonasm
- https://github.com/Flamarine/Flyx
- https://github.com/LemmaEOF/ItsPoppin/tree/trunk
- https://github.com/unilock/motd
- https://github.com/unilock/MouseFix
- https://git.sleeping.town/Nil/NarratorShutUp
- https://github.com/FoundationGames/NilChasm
- https://github.com/unilock/nilcord/tree/nilloader/1.4.7 / https://modrinth.com/mod/nilcord/version/2.1.0+1.4.7+nilloader
- https://git.sleeping.town/Rewind/NiteLoader
- https://git.sleeping.town/Nil/NullifyTelemetry
- https://github.com/jaskarth/retromods/tree/main/147/overworld_2
- https://git.sleeping.town/Rewind/retroEMI
- https://git.sleeping.town/Nil/SHA1Redemption
- https://git.sleeping.town/Rewind/UpsilonFixes
- https://git.gay/j0/ZeroBridge

(that might be all of them?)

**Any additional context?:** I previously opened an issue about NilLoader before noticing this thread: https://github.com/modrinth/labrinth/issues/903

**Comment by OracleHisty** [comment]
**What loader do you suggest?:** BEDROCK

**Please provide a link to the loader (website, GitHub, etc.):** https://www.minecraft.net

**When, approximately, was this loader created?:** Sunsetting of Legacy Console Edition

**What project type(s) would this loader be applicable to?:** 
> Resource Packs (official compat outside of the minecraft marketplace)
> Addons ([See This For Context](https://github.com/orgs/modrinth/discussions/105#discussioncomment-11792604)
> A hypothetical Maps Category

**Please provide links to at least three projects that use this loader:**
- Unfortunately there is none that currently use it outside of CurseForge or MCPEdl, however here are a few accidentally deceptive results on CurseForge due to their legacy formatting being totally un-planned:
[Curseforge's Addon's Category](https://www.curseforge.com/minecraft/search?page=1&pageSize=20&sortBy=relevancy&class=mc-addons)

It is commented on curseforge that mcaddons are intended for bedrock, but it was a poorly planned category with stereotypical curseforge auto-moderation and lack of quality-control.

**Any additional context?:** For the sake of futureproofing and saving on web traffic congestion with inevitable confusion.
